pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -168,29 +368,91 @@ func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, p
+			return &status, err
 		}
 	}
 
 	// At this point, all of the current Replicas are Running and Ready, we can consider termination.
 	// We will wait for all predecessors to be Running and Ready prior to attempting a deletion.
 	// We will terminate Pods in a monotonically decreasing order over [len(pods),set.Spec.Replicas).
-	// Note that we do not resurrect Pods in this interval.
+	// Note that we do not resurrect Pods in this interval. Also not that scaling will take precedence over
+	// updates.
 	if unhealthy > 0 && monotonic {
 		glog.V(2).Infof("StatefulSet %s is waiting on %d Pods", set.Name, unhealthy)
-		return nil
+		return &status, nil
 	}
 	for target := len(condemned) - 1; target >= 0; target-- {
 		glog.V(2).Infof("StatefulSet %s terminating Pod %s", set.Name, condemned[target])
 		if err := ssc.podControl.DeleteStatefulPod(set, condemned[target]); err != nil {
-			return err
+			return &status, err
+		}
+		if getPodRevision(pods[target]) == currentRevision.Name {
+			status.CurrentReplicas--
+		} else if getPodRevision(pods[target]) == updateRevision.Name {
+			status.UpdatedReplicas--
 		}
 		if monotonic {
-			return nil
+			return &status, nil
+		}
+	}
+
+	// for the OnDelete strategy we short circuit. Pods will be updated when they are manually deleted.
+	if set.Spec.UpdateStrategy.Type == apps.OnDeleteStatefulSetStrategyType {
+		return &status, nil
+	}
+
+	// we compute the minimum ordinal of the target sequence for a destructive update based on the strategy.
+	updateMin := 0
+	if set.Spec.UpdateStrategy.Type == apps.PartitionStatefulSetStrategyType {
+		updateMin = int(set.Spec.UpdateStrategy.Partition.Ordinal)
+	}
+	// we terminate the Pod with the largest ordinal that does not match the update revision.
+	for target := len(replicas) - 1; target >= updateMin; target-- {
+		if getPodRevision(replicas[target]) != updateRevision.Name {
+			glog.V(2).Infof("StatefulSet %s terminating Pod %s for update", set.Name, replicas[target])
+			err := ssc.podControl.DeleteStatefulPod(set, replicas[target])
+			status.CurrentReplicas--
+			return &status, err
+		}
+	}
+	return &status, nil
+}
+
+// updateStatefulSetStatus updates set's Status to be equal to status. If status indicates a complete update, it is
+// mutated to indicate completion. If status is semantically equivalent to set's Status no update is performed. If the
+// returned error is nil, the update is successful.
+func (ssc *defaultStatefulSetControl) updateStatefulSetStatus(
+	set *apps.StatefulSet,
+	status *apps.StatefulSetStatus) error {
+
+	if set.Spec.UpdateStrategy.Type == apps.RollingUpdateStatefulSetStrategyType &&
+		status.UpdatedReplicas == status.Replicas &&
+		status.ReadyReplicas == status.Replicas {
+		status.CurrentReplicas = status.UpdatedReplicas
+		status.CurrentRevision = status.UpdateRevision
+		status.UpdateRevision = ""
+		status.UpdatedReplicas = 0
+	}
+
+	if set.Status.ObservedGeneration == nil ||
```

Comment:
Invert this condition to drop a level of indentation.  Also, these are unreadable - you may just want to split the tests into small methods.